File: rs/hang/src/catalog/compression.rs
use crate::Error;

use derive_more::Display;
use std::str::FromStr;

/// The compression applied to a metadata track's frames.
///
/// A track's catalog entry declares this explicitly; the conventional `.z` suffix on a track name
/// is a naming convention, not a signal a consumer may rely on. An absent field (`None` on the
/// config) means the frames are uncompressed.
#[derive(Debug, Clone, PartialEq, Eq, Display)]
#[non_exhaustive]
pub enum Compression {
	/// Group-scoped raw DEFLATE ([RFC 1951](https://www.rfc-editor.org/rfc/rfc1951.html)),
	/// sync-flushed at each frame boundary so every frame is self-delimited while later frames
	/// compress against the earlier ones in the same group.
	#[display("deflate")]
	Deflate,

	/// A compression this build does not recognize, preserved verbatim.
	///
	/// A consumer MUST ignore the track, since its frames cannot be decoded.
	#[display("{_0}")]
	Unknown(String),
}

impl FromStr for Compression {
	type Err = Error;

	fn from_str(s: &str) -> Result<Self, Self::Err> {
		Ok(match s {
			"deflate" => Self::Deflate,
			_ => Self::Unknown(s.to_string()),
		})
	}
}

#[cfg(test)]
mod test {
	use super::*;

	#[test]
	fn deflate_roundtrip() {
		let mut config = crate::catalog::JsonConfig::new(crate::catalog::Mode::Stream);
		config.compression = Some(Compression::Deflate);
		let encoded = serde_json::to_value(&config).unwrap();
		assert_eq!(encoded["compression"], serde_json::json!("deflate"));
		assert_eq!(
			serde_json::from_value::<crate::catalog::JsonConfig>(encoded)
				.unwrap()
				.compression,
			Some(Compression::Deflate)
		);
	}

	#[test]
	fn unknown_roundtrip() {
		let wire = serde_json::json!({ "mode": "stream", "compression": "zstd" });
		let config: crate::catalog::JsonConfig = serde_json::from_value(wire.clone()).unwrap();
		assert_eq!(config.compression, Some(Compression::Unknown("zstd".to_string())));
		assert_eq!(serde_json::to_value(&config).unwrap(), wire);
	}
}

File: js/hang/src/catalog/compression.ts
import * as z from "zod/mini";

/** The compression algorithms this build knows how to decode. */
export type KnownCompression = "deflate";

/**
 * The compression applied to a data track's frames, or absent when they are uncompressed.
 *
 * `"deflate"` is group-scoped raw DEFLATE (RFC 1951), sync-flushed at each frame boundary: every
 * frame is self-delimited while later frames compress against the earlier ones in the same group.
 *
 * A track's catalog entry declares this explicitly. The `.z` suffix seen on some track names is a
 * naming convention, not a signal: never infer compression from a track's name.
 *
 * Typed as a plain string so an unrecognized algorithm survives a reparse-and-republish verbatim.
 * Narrow it with {@link compressionSupported} before reading the track, and ignore the track
 * otherwise, since its frames cannot be decoded.
 */
export const CompressionSchema = z.string();

/** The compression applied to a data track's frames. See {@link CompressionSchema}. */
export type Compression = z.infer<typeof CompressionSchema>;

/** Whether this build knows how to decode frames compressed with `compression`. */
export function compressionSupported(compression: Compression): compression is KnownCompression {
	return compression === "deflate";
}
